### DISCUSSION TITLE
RUSTC_WRAPPER equivalent

### DISCUSSION THREAD

**Discussion by shabnam4b** [bazelbuild/rules_rust#1852] [Q&A]
Hi folks, 

I'd like to use a rustc wrapper (RUSTC_WRAPPER equivalent) but I could not figure how to do it. 
I think one solution is being able to use rust toolchain from my local machine and I can replace rustc with the wrapper. 

Please advise, thanks!

**Comment by UebelAndre** [comment]
This should mostly be doable by updating the `rust_toolchain` to point to whatever file you want. Though `rust_toolchain` doesn't have a place to add the original rustc. Maybe the toolchain can be updated to consume the runfiles of whatever is passed to `rust_toolchain::rustc` if there are any and include them in it's `all_files` member?

**Reply by shabnam4b** [reply]
Can you please give me an example?

One other approach is to extract all the rustc commands during the bazel build and replay them. It's error-prone as I could not see the actual rustc logs using `--subcommand `and I need to write a good parser but it's doable I guess.

**Reply by UebelAndre** [reply]
What I'm saying is something like this is not quite possible without making modifications to `rules_rust` to support it. I don't think something specifically like `RUSTC_WRAPPER` would be supported though since that would make actions non-hermetic. My current thought is to update https://github.com/bazelbuild/rules_rust/blob/0.18.0/rust/toolchain.bzl#L420-L428 to extract runfiles from `rustc` and pass them to `all_files`. @krasimirgg what do you think about that?

**Reply by shabnam4b** [reply]
Oh, ok. I was looking for a way to use a wrapper without modifying the current rule. Could I create an issue for supporting rustc_wrapper? Regardless of the implementation details, I think it is useful to consider supporting it. There are some program analysis tools that rely on it. I have not checked bazel codes yet but if you folks think it can become a widely used feature, I can work on it. Otherwise, I prefer to find a workaround for now :)

**Comment by hermanventer** [comment]
Rather than trying to make Bazel more like Cargo, perhaps it would make sense to just write Bazel rules for MIRAI? Since there will be a lot of commonality between Rustc and MIRAI rules, this project is probably the right place to do it.

**Comment by shabnam4b** [comment]
Hi Herman,

I'm not sure how building MIRAI with bazel resolves the issue, as MIRAI
should be used instead of rustc in building other projects. The hermetic
build requirement of bazel makes it impossible to use the rustc binary from
my own local machine. One solution might be to write a new toolchain. In
general, the ultimate solution would be for rust bazel rules to support
rust_wrapper.

For now, I have used a workaround (although not a perfect one): I modified
the tool_path
<https://github.com/bazelbuild/rules_rust/blob/main/rust/private/rustc.bzl#L843>in
the private rust rules so that it points to MIRAI instead of rustc.

Best,

 Shabnam
